### PR TITLE
Allow to add custom mailbox roles on initilizing client

### PR DESCRIPTION
**Pull request by RockyMeadow**

### CODE REVIEW

File: lib/client/Client.js [hunk 1]
```diff
@@ -126,6 +126,22 @@ export default class Client {
...
+   * Initializes the client with a list of custom mailbox roles.
+   * Mailboxes having these custom roles will be preserved their role.
+   * 
+   * @param roles {Array} The list of custom roles to be added
+   * 
+   * @return {Client} This Client instance.
+   */
+  withCustomMailboxRoles(roles) {
+    roles.forEach(role => {
+      MailboxRole[role.toUpperCase()] = new MailboxRole(role);
```
Comment (chibenwa) [inline lib/client/Client.js]: Generally upon case operation in Java we always pass a locale.

For example take the example of the infamous turkish `i`. A toLowerCase on `I` will not give you the same result on US and turkish locales.

There many more subtile differences.

See http://www.informit.com/articles/article.aspx?p=1751371&seqNum=12 for more details.... There should be some JS equivalent.

File: lib/client/Client.js [hunk 2]
```diff
@@ -126,6 +126,22 @@ export default class Client {
     return this;
   }
 
+  /**
+   * Initializes the client with a list of custom mailbox roles.
```
Comment (tuanlc) [inline lib/client/Client.js]: Use `<br />` if you expect a new line after. Check other function docs.

File: lib/client/Client.js [hunk 3]
```diff
@@ -126,6 +126,22 @@ export default class Client {
...
+   * Mailboxes having these custom roles will be preserved their role.
+   * 
+   * @param roles {Array} The list of custom roles to be added
+   * 
+   * @return {Client} This Client instance.
+   */
+  withCustomMailboxRoles(roles) {
```
Comment (tuanlc) [inline lib/client/Client.js]: should we support default: `roles = []`?

File: lib/client/Client.js (hunk 1 above)
Comment (RockyMeadow) [reply lib/client/Client.js]: Unfortunately, there is no JS equivalent.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toUpperCase

Comment (chibenwa) [reply lib/client/Client.js]: Héhé so this mean I have one more :trollface: about JS...

**Review by tuanlc — APPROVED** [review]

File: lib/client/Client.js (hunk 1 above)
Comment (mbaechler) [reply lib/client/Client.js]: Well, you didn't read to the end of the page, there's a link to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleUpperCase

File: lib/client/Client.js [hunk 4]
```diff
@@ -126,6 +126,22 @@ export default class Client {
     return this;
   }
 
+  /**
+   * Initializes the client with a list of custom mailbox roles.<br />
+   * Mailboxes having these custom roles will be preserved their role.
```
Comment (mbaechler) [inline lib/client/Client.js]: s/Mailboxes having these custom roles will be preserved their role./Mailboxes having these custom roles will preserve their role./

File: lib/client/Client.js [hunk 5]
```diff
@@ -126,6 +126,22 @@ export default class Client {
...
+   *
+   * @param roles {Array} The list of custom roles to be added
```
Comment (mbaechler) [inline lib/client/Client.js]: Array of what type ?

File: test/common/Client.js [hunk 6]
```diff
@@ -130,6 +130,14 @@ describe('The Client class', function() {
 
   });
 
+  describe('The withCustomMailboxRoles method', function() {
+    it('should allow custom role mailbox after initializing client with custom role', function() {
```
Comment (mbaechler) [inline test/common/Client.js]: what's the effect of defining these roles? I could make this test pass with a method doing `return this`.
You miss some tests about the behavior it adds to the mailboxes.

**Review by chibenwa — COMMENTED** [review]
useless comment sorry

File: test/common/Client.js (hunk 6 above)
Comment (chibenwa) [reply test/common/Client.js]: We can, for example, assert that this is equal to the mailboxRole object we expect?

File: lib/client/Client.js (hunk 1 above)
Comment (RockyMeadow) [reply lib/client/Client.js]: Thanks, i used  toUpperCase with en-US as locale

File: test/common/Client.js (hunk 6 above)
Comment (RockyMeadow) [reply test/common/Client.js]: I changed to expecting Mailbox role to be a certain value after doing a `getMailboxes`

File: test/common/Client.js [hunk 7]
```diff
@@ -130,6 +130,28 @@ describe('The Client class', function() {
 
   });
 
+  describe('The withCustomMailboxRoles method', function() {
+    it('should add a list of mailbox roles to be preserved', function(done) {
+      const mailboxes = [
+        { id: 'id', name: 'test1', role: 'for-test' },
+        { id: 'id', name: 'test2', role: 'invalid' }
```
Comment (chibenwa) [inline test/common/Client.js]: I don't think it's invalid.

I think it's just `non-custom-mailbox` right?

Comment (chibenwa) [inline test/common/Client.js]: I don't think it's invalid.

I think it's just `non-custom-mailbox` right?

**Review by chibenwa — APPROVED** [review]

File: test/common/Client.js [hunk 8]
```diff
@@ -130,6 +130,28 @@ describe('The Client class', function() {
...
+        { id: 'id', name: 'test2', role: 'non-custom-mailbox' }
+      ];
+
+      new jmap.Client({
+        post: () => q([['mailboxes', { list: mailboxes }, '#0']])
+      })
+        .withCustomMailboxRoles(['for-test'])
+        .getMailboxes()
+        .then(mailboxes => {
+          expect(mailboxes[0].role.value).to.equal('for-test');
```
Comment (mbaechler) [inline test/common/Client.js:146]: is the ordering guaranteed?

Comment (RockyMeadow) [reply test/common/Client.js:146]: Yes

**Pull request merged**